kernel/src/process.rs

```diff
@@ -1078,9 +1081,9 @@ pub struct FunctionCall {
     /// The third argument to the function.
     pub argument2: usize,
     /// The fourth argument to the function.
-    pub argument3: usize,
+    pub argument3: CapabilityPtr,
```

Comment:
This is where we pass back the userdata (a `void*`) passed in via `subscribe`.

Reply:
Will add comment as to why

Reply:
This is a generic type, where happen to not use the other arguments for capabilities returned to userspace. However, for the sake of consistency, shouldn't these really not all be `CapabilityPtr` arguments?

I'd rather get this right related to the semantics of this type (a thing that schedules an upcall with a given set of argument), instead of getting it working just for the current use cases with our (at this stage of the code) arbitrary convention of passing a userspace pointer in the fourth argument.

Reply:
A function pointer, three integers, and a data pointer seems semantically correct to me for an upcall. Is `FunctionCall` used for anything other than upcalls? If not, this seems fine to me (modulo the `usize` versus `u32` discussion).

Reply:
I agree that this is correct for an upcall, but I think that `FunctionCall` is supposed to be a more generic type. For instance, it is also used to set the `_start` function of a process:

https://github.com/tock/tock/blob/8836ded912c263ccdbb704b615ae410e61c610fb/kernel/src/syscall.rs#L748-L751

https://github.com/tock/tock/blob/8836ded912c263ccdbb704b615ae410e61c610fb/kernel/src/syscall.rs#L782-L788

https://github.com/tock/tock/blob/8836ded912c263ccdbb704b615ae410e61c610fb/kernel/src/process_standard.rs#L1964-L1973

Reply:
As @jrvanwhy said, the types chosen were to match userspace.

However, I do see that second use @lschuermann, which I agree is slightly different than a regular upcall.

But given that the startup ABI passes `usize` or `u32` for those arguments, I don't think we have anything that motivates promoting those arguments. And there is an actual cost on CHERI to making them wider.

@bradjc this argument corresponds to the only pointer in the subscribe interface, and because the user intends to dereference it (even though the kernel does not), it should carry authority.

Reply:
> I don't understand why this would be a pointer since it can be any register sized thing.

I think this, again, boils down to the confusion that `CapabilityPtr` can hold either any register value, or a pointer with metadata. Because `FunctionCall` is supposed to be able to convey any type of argument to userspace, we'll need to be able to pass either. For the sake of consistency, I thus argue to make all the arguments into a `CapabilityPtr`.

I agree that this type name is unintuitive in this context, but we can't seem to be able to come up with anything better.

Reply:
Let's leave it like this for now, and perhaps extend this to be all `CapabilityPtr` in a follow-up PR.